premerge/post-submit-testing.md

```diff
@@ -0,0 +1,202 @@
+be produced if certain projects (that were only triggered by some files) were
+failing and some were passing which would significantly increase false positive
+notifications. For example, supposed that we have three commits that land in
+`main` and run through postcommit testing: commit A that touches MLIR, commit B
+that touches clang-tidy, and commit C that modifies MLIR. Commit A lands, then
+commit B, then commit C. If commit A introduces MLIR test failures into an
+otherwise clean slate, we would see the following events:
+
+1. Commit A lands. Because it touches MLIR, the buildbot worker runs the MLIR
+   tests. Some of the tests fail. The buildbot "turns red" and a notification is
+   sent out to the PR author.
+2. Commit B lands. Since it touches clang-tidy, the buildbot worker runs the
+   clang-tidy tests. All of the tests pass. The buildbot "turns green". No
+   notifications are sent out since everything is passing.
+3. Commit C lands. Since it touches MLIR, the buildbot workers runs the MLIR
+   tests. The problem introduced in commit A still exists, so some tests fail.
+   No new tests fail. Since the buildbot was previously green due to the
+   interspersed clang-tidy commit, a notification is still sent out to the
+   author of commit C.
+
+By running the tests for all projects in every postsubmit test run, we avoid
+the problematic situation described above.
+
+Another reason for running all the tests in every postsubmit run: When running
+premerge tests on a PR, we also explicitly do not test certain projects even
+though their dependencies change. While we do this because we suspect
+interactions resulting in test failures would be quite rare, it is possible, and
+having a postcommit configuration catch these rare failures would be useful.
+
+### Data Storage
+
+The hosted Buildbot master instance at [lab.llvm.org](https://lab.llvm.org)
+contains results for all recent postcommit runs. We plan on querying the results
+from the buildbot master because they are already available and that is where
+they will natively be reported after the infrastructure is set up. Buildbot
+supports a [REST API](https://docs.buildbot.net/latest/developer/rest.html) that
+would allow for easily querying the state of a commit in `main`.
+
+For the proposed premerge advisor that tells the user what tests/build failures
+they can safely ignore, we need to know what is currently failing on `main`.
+Each pull request is tested as if it was merged into main, which means the
+commit underneath the PR is very recent. If a premerge run fails, the premerge
+advisor will find the commit from `main` the PR is being tested on. It will then
+query the Buildbot master using the REST API for the status of that commit.
+It can then report the appropriate status to the user. Having the status will
+let the premerge advisor avoid pestering LLVM developers with failures
+unrelated to their changes.
```

Comment:
What happens if the PR build finished & fails before the main post-commit build finished? Is there a queue/poll mechanism for the advisor? Or will it report "don't know" ?

Reply:
The plan is to just have it query for the preceding commits until it hits something. For the purpose of excluding test failures at head, that should be fine.

I'll flesh out the details more for this when I write up a document/RFC on the premerge advisor specifically.